### ISSUE TITLE
Should the sections be reordered?

### ISSUE THREAD

**Issue by lgolding** [oasis-tcs/sarif-spec#278]
@kupsch asked about the section ordering in the spec (presumably because he couldn’t make heads or tails of the ordering I’ve chosen). Is there an ISO recommendation for this?

What I’ve done is:

* First come generally applicable sections (like the description of date-time-valued properties), and the descriptions of objects that occur widely throughout the spec (like `message` objects and `fileLocation` objects).
* Then come sections for each object, generally in top-down order (so `sarifLog` comes first, `run` comes later, `result` later still, then `location`, `physicalLocation`, …). But when an object has children of two peer types (for example, a `physicalLocation` has both `fileLocation` and `region`), I arbitrarily put the one I think is most useful first.
* With the section for each object, I generally put what I consider the most useful properties first, and I generally group related properties together (such as `invocation.stdout`, `stderr`, `stdin`, and `stdoutStderr`).

All this ends up looking pretty arbitrary and confusing, but I think it would be even more confusing to (for example) order the objects and the properties within each object purely alphabetically.

I also worry that at this point, a wholesale reordering would change-bar the entire document, making it impossible for reviewers to see the changes (which are already large).

**Comment by michaelcfanning** [comment]
Your ordering is sensible. If you have thoughtfully grouped properties in a useful order, I wouldn't disturb that (in fact, might double-down and make another pass for this value). The reason is that any IDE or other Intellisense generating thing can always alphabetize if they care to but your thoughtful ordering can't be recovered from an alphabetized list. An alphabetized order (in the spec in particular) will also be prone to situations where a user uncovers a specific property and wonders whether some sensible matching concept (which begins with a different letter) has been defined. 

I notice that ordering tends to be presevered in various schema processing things, which we can take advantage of even though not guaranteed. When we change the position of properties in the schema, for example, this has the practical effect of repositioning it in serialized form (because our OM generator is order sensitive and JSON.NET is reflection order sensitive). Note that our code hints file (that helps tune the SARIF OM generation) is strictly alphabetized, which makes sense given its purpose).

If we reject the idea of alphabetizing object properties, are there other changes we could/should consider?

Also, contingent on time available, we could scrub the schema to make sure it is ordered identically with the spec, to ensure we're realizing the value you've invested in the document.

**Comment by lgolding** [comment]
Thanks @michaelcfanning. @kupsch, I am going to close this `resolved-by-design`.

**Comment by dmk42** [comment]
@lgolding - I've kept this in the back of my mind during my interactions with ISO so I could give a detailed response.

ISO prefers that standards be organized along the lines of the relationships between its parts, as you have done, rather than alphabetically.  The reason is that the national standards bodies of various countries translate standards into other languages, and it is easiest to check that a translation is correct if it follows the exact sequence of the original.  If the standard is alphabetized, it will be in a different order in a new language.

In the past, the conflict between the desire for a logical organization and the need to find things quickly was resolved by having an index, the one place where things are alphabetized.  However, in recent years, electronic document searches have removed the need for indexes, and some people at ISO HQ are even starting to consider discouraging indexes because, as a standard gets transcribed from one electronic format to another, the meaning of an index reference might be difficult to translate (e.g. the fact that a term is defined on page 67 has no meaning in a format without pages).

So, what they prefer is exactly what you have.
